vic/drivers/shared_image/src/vic_mpi_support.c

```diff
@@ -1671,6 +1453,179 @@ create_MPI_param_struct_type(MPI_Datatype *mpi_type)
+        log_err("MPI error: %d\n", status);
+    }
+
+    status = MPI_Type_commit(mpi_type);
+    if (status != MPI_SUCCESS) {
+        log_err("MPI error: %d\n", status);
+    }
+
+    // cleanup
+    free(blocklengths);
+    free(offsets);
+    free(mpi_types);
+}
+
+/******************************************************************************
+ * @brief   Create an MPI_Datatype that represents the alarm_struct
+ * @details This allows MPI operations in which the entire alarm_struct
+ *          can be treated as an MPI_Datatype.
+ * @param mpi_type MPI_Datatype that can be used in MPI operations
+ *****************************************************************************/
+void
+create_MPI_alarm_struct_type(MPI_Datatype *mpi_type)
+{
+    int           nitems; // number of elements in struct
+    int           status;
+    int          *blocklengths;
+    size_t        i;
+    MPI_Aint     *offsets;
+    MPI_Datatype *mpi_types;
+    MPI_Datatype  mpi_dmy_type;
+
+    // nitems has to equal the number of elements in alarm_struct
+    nitems = 6;
+    blocklengths = malloc(nitems * sizeof(*blocklengths));
+    if (blocklengths == NULL) {
+        log_err("Memory allocation error")
+    }
+
+    offsets = malloc(nitems * sizeof(*offsets));
+    if (offsets == NULL) {
+        log_err("Memory allocation error")
+    }
+
+    mpi_types = malloc(nitems * sizeof(*mpi_types));
+    if (mpi_types == NULL) {
+        log_err("Memory allocation error")
+    }
+
+    // none of the elements in location_struct are arrays.
+    for (i = 0; i < (size_t) nitems; i++) {
+        blocklengths[i] = 1;
+    }
+
+    // reset i
+    i = 0;
+
+    // unsigned int count;
+    offsets[i] = offsetof(alarm_struct, count);
+    mpi_types[i++] = MPI_UNSIGNED;
+
+    // int next;
+    offsets[i] = offsetof(alarm_struct, next);
+    mpi_types[i++] = MPI_INT;
+
+    // unsigned int freq;
+    offsets[i] = offsetof(alarm_struct, freq);
+    mpi_types[i++] = MPI_UNSIGNED;
+
+    // int n;
+    offsets[i] = offsetof(alarm_struct, n);
+    mpi_types[i++] = MPI_INT;
+
+    // dmy_struct date;
+    offsets[i] = offsetof(alarm_struct, date);
+    create_MPI_dmy_struct_type(&mpi_dmy_type);
+    mpi_types[i++] = mpi_dmy_type;
+
+    // bool is_subdaily;
+    offsets[i] = offsetof(alarm_struct, is_subdaily);
+    mpi_types[i++] = MPI_C_BOOL;
+
+    // make sure that the we have the right number of elements
+    if (i != (size_t) nitems) {
+        log_err("Miscount: %zd not equal to %d\n", i, nitems);
+    }
+
+    status = MPI_Type_create_struct(nitems, blocklengths, offsets, mpi_types,
+                                    mpi_type);
+    if (status != MPI_SUCCESS) {
+        log_err("MPI error: %d\n", status);
+    }
+
+    status = MPI_Type_commit(mpi_type);
+    if (status != MPI_SUCCESS) {
+        log_err("MPI error: %d\n", status);
+    }
+
+    // cleanup
+    free(blocklengths);
+    free(offsets);
+    free(mpi_types);
+    MPI_Type_free(&mpi_dmy_type);
```

Comment:
see comment above. there is allocated memory inside the `mpi_dmy_type` structure.